### ISSUE TITLE
Runtime crash when using `any` with a type alias `any SomeProtocol & SomeClass`

### ISSUE THREAD

**Issue by amomchilov** [swiftlang/swift#59566]
**Describe the bug**

It's hard to summarize concisely, but in short stacking `any` twice (e.g. via type alias) with a protocol composition with a class causes a runtime crash.

**To Reproduce**

```swift
protocol P {
	func f()
}

protocol AnotherProtocol {}

class SuperClass {}

class Subclass: SuperClass, P, AnotherProtocol  {
	func f() { print("S.f()") }
}

typealias AnyPAndSuperClass = any P & SuperClass

// `any` used with a typealias that itself uses `any` and protocol composition with a class causes a crash
//(Subclass() as any (any P & SuperClass)).f() // 💥 Terminated due to signal: BUS ERROR (10)
// The typealias isn't actually necessary, inlining has the same issue
//(Subclass() as any (any P & SuperClass)).f() // 💥 Terminated due to signal: BUS ERROR (10)

// Omitting the "outter" `any` works fine
(Subclass() as AnyPAndSuperClass).f() // ✅

// If the protocol composition doesn't contain a class, it seems to work fine
typealias AnyPAndUnecessaryProtocol = any P & AnotherProtocol
(Subclass() as any AnyPAndUnecessaryProtocol).f() // ✅
(Subclass() as AnyPAndUnecessaryProtocol).f() // ✅
```

**Expected behavior**

```
Terminated due to signal: BUS ERROR (10)
```

**Environment (please complete the following information):**
 - OS: macOS Ventura `13.0 Beta (22A5266r)`
 - Xcode Version/Tag/Branch: `Version 14.0 beta (14A5228q)`

**Comment by CodaFi** [comment]
This code shouldn't survive to runtime. ToT says:

```
<source>:16:16: error: redundant 'any' has no effect on existential type 'any SuperClass & P'
(Subclass() as any (any P & SuperClass)).f() // 💥 Terminated due to signal: BUS ERROR (10)
<source>:18:16: error: redundant 'any' has no effect on existential type 'any SuperClass & P'
(Subclass() as any (any P & SuperClass)).f() // 💥 Terminated due to signal: BUS ERROR (10)
```

@hborla Does this look familiar.

**Comment by amomchilov** [comment]
Hey Robert, I really enjoyed your podcast with John Sundell!

This compiled without warning for me both from a standalone file (`swift /Users/Alex/Desktop/crasher.swift`) and when pasted in a `swift repl`. I extracted this minimal example from a full Xcode project, but in that case, I actually got a compile-time ambiguous expression error instead (that I haven't been able to replicate in a simple example). I posted some details about it [on this forum post](https://forums.swift.org/t/compiler-bug-adding-explicit-any-to-a-protocol-compositon-type-breaks-the-callers/58253). Not sure if that warrants its own separate bug report or not.

Semi-related question: Isn't the `any` in `typealias PandC = any SomeProtocolP & SomeClass` always redundant?

As far as I understand it, the existential containers that values get boxed into are necessary because values don't store their own type information (no `isa` pointer), and they're heterogeneously sized. Instances of classes have neither of these problems (each object has an `isa`, and each reference is the same size). Thus, no existential boxing would be necessary for `PandC`, and the purpose of explicit `any` (drawing attention to potentially costly boxing operations) doesn't apply. Is that right?
